src/basic/alloc-util.h

```diff
@@ -60,6 +60,16 @@ typedef void* (*mfree_func_t)(void *p);
                 0;                              \
         })
 
+#define free_and_replace_fn(a, b, free_fn)  \
+        ({                                      \
+                typeof(a)* _a = &(a);           \
+                typeof(b)* _b = &(b);           \
+                free_fn(*_a);                   \
+                *_a = *_b;                      \
+                *_b = NULL;                     \
+                0;                              \
+        })
```

Comment:
I actually implemented this in https://github.com/systemd/systemd/pull/24067 already, if you rebase on main you can use `free_and_replace_full`